apps/sms/services/js/message_manager.js

```diff
@@ -429,26 +433,32 @@ var MessageManager = {
     // other potential call to the API, like the one for getting a message
     // list, could be done within the calls to mark the messages as read.
 
-    var id = list.pop();
-    // TODO: Third parameter of markMessageRead is return read request.
-    //       Here we always return read request for now, but we can let user
-    //       decide to return request or not in Bug 971658.
-    var req = this._mozMobileMessage.markMessageRead(id, isRead, true);
-    // isRead == false i.e mark as unread case, marking only one message
-    // as unread is sufficient.
-    req.onsuccess = (function onsuccess() {
-      if (!list.length || !isRead) {
-        return;
-      } else if (isRead) {
-        this.markMessagesRead(list, isRead);
-      }
-    }).bind(this);
+    // sendReadReport == true if Send-Read-Reports switch in Messaging Settings
+    // is enabled else false
+    Settings.sendReadReport().then((result) => {
+      console.log(result);
+      markMessage(result);
+    }, (err) => {
+      console.error(err);
+    });
 
-    req.onerror = function onerror() {
-      console.error(
-        'Error while marking message %d as read: %s', id, this.error.name
-      );
-    };
+    function markMessage(sendReadReport) {
+      var id = list.pop();
+      var req = self._mozMobileMessage.markMessageRead(id,
+        isRead, sendReadReport);
+
+      req.then(() => {
+        // if isRead == false then mark only one message in each thread
+        if (!list.length || !isRead) {
+          deferred.resolve();
+          return deferred.promise;
+        } else {
+          markMessage(sendReadReport);
+        }
+      }, () => {
+        console.log('ERROR in mozMobileMessage.markMessageRead method');
+      });
+    }
```

Comment:
I think you can do simpler, without using a deferred. Deferred are useful when we don't have promises in the APIs we use, but here we have promises (or then-able which is essentially the same).

How about something like this:
```javascript
if (!this._mozMobileMessage) {
  return Promise.resolve();
}

var markMessage = (shouldSendReadReport) => {
  if (!list.length) {
    return Promise.resolve();
  }

  var id = list.pop();
  var req = self._mozMobileMessage.markMessageRead(id, isRead, shouldSendReadReport);
  return req.then(
    () => isRead ? markMessage(shouldSendReadReport) : Promise.resolve()
  );
};

return Settings.sendReadReport.then(markMessage);
```

If you don't understand porperly, remember the promises are "transitive" (not sure it's the right word, hope you'll understand it).